src/runmode-dpdk.c

```diff
@@ -0,0 +1,1278 @@
+/* Copyright (C) 2021 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \ingroup dpdk
+ *
+ * @{
+ */
+
+/**
+ * \file
+ *
+ * \author Lukas Sismis <lukas.sismis@gmail.com>
+ *
+ * DPDK runmode
+ *
+ */
+
+#include "suricata-common.h"
+#include "runmodes.h"
+#include "runmode-dpdk.h"
+#include "source-dpdk.h"
+#include "util-runmodes.h"
+#include "util-byte.h"
+
+#ifdef HAVE_DPDK
+#include <rte_log.h>
+#include <rte_mempool.h>
+#include <rte_eal.h>
+#include <rte_lcore.h>
+
+#define RSS_HKEY_LEN 40
+static uint8_t rss_hkey[] = { 0x6D, 0x5A, 0x6D, 0x5A, 0x6D, 0x5A, 0x6D, 0x5A, 0x6D, 0x5A, 0x6D,
```

Comment:
What is this hardcoded value ?

Reply:
This is the haskey for RSS symmetric hashing. That is also used in our guides for high performance, see https://suricata.readthedocs.io/en/latest/performance/packet-capture.html

Reply:
This value is a special hash key that configures receive side scaling in a way that it puts packets of one flow (but from both directions) into one NIC queue. Packets are symmetrically scattered to NIC queues using a hash function. 
Suricata workers then fetch packets from these NIC queues for further processing. Configuring RSS with this hash key ensures packets of one flow land on the same worker every time, thus not requiring multi-thread communication. Other hash keys usually do not have this property.

Other receive runmodes (e.g. AF_PACKET) take the configured NIC and interact with them without any configuration. However, in DPDK each application that uses a NIC is required to configure the NIC by itself. For that reason, it was necessary to implement also the configuration phase (you would normally do that with `ethtool` prior to running Suricata). 

More on the RSS hash key value: https://www.ndsl.kaist.edu/~kyoungsoo/papers/TR-symRSS.pdf

Reply:
@lukashino in port_conf you are assigning RTE_IP_RSS, but in the hash key you have defines suing layer 4. there is conflict in what is been explained and used, Please rework

Reply:
Thanks, I've changed it to ETH_RSS_TCP | ETH_RSS_UDP
If I am not mistaken, this should be correct.